llm/ggml_llama.go

```diff
@@ -0,0 +1,674 @@
+		return "Q3_K_S"
+	case llamaFileTypeQ3_K_M:
+		return "Q3_K_M"
+	case llamaFileTypeQ3_K_L:
+		return "Q3_K_L"
+	case llamaFileTypeQ4_K_S:
+		return "Q4_K_S"
+	case llamaFileTypeQ4_K_M:
+		return "Q4_K_M"
+	case llamaFileTypeQ5_K_S:
+		return "Q5_K_S"
+	case llamaFileTypeQ5_K_M:
+		return "Q5_K_M"
+	case llamaFileTypeQ6_K:
+		return "Q6_K"
+	default:
+		return "Unknown"
+	}
+}
+
+type Running struct {
+	Port int
+	Cmd  *exec.Cmd
+}
+
+type llama struct {
+	api.Options
+	Running
+}
+
+func newLlama(model string, adapters []string, opts api.Options) (*llama, error) {
+	if _, err := os.Stat(model); err != nil {
+		return nil, err
+	}
+
+	if _, err := os.Stat(runner); err != nil {
+		return nil, err
+	}
+
+	if len(adapters) > 1 {
+		return nil, errors.New("ollama supports only one lora adapter, but multiple were provided")
+	}
+
+	params := []string{
+		"--model", model,
+		"--ctx-size", fmt.Sprintf("%d", opts.NumCtx),
+		"--gqa", fmt.Sprintf("%d", opts.NumGQA),
+		"--rope-freq-base", fmt.Sprintf("%f", opts.RopeFrequencyBase),
+		"--rope-freq-scale", fmt.Sprintf("%f", opts.RopeFrequencyScale),
+		"--batch-size", fmt.Sprintf("%d", opts.NumBatch),
+		"--n-gpu-layers", fmt.Sprintf("%d", opts.NumGPU),
+		"--embedding",
+	}
+
+	if len(adapters) > 0 {
+		// TODO: applying multiple adapters is not supported by the llama.cpp server yet
+		params = append(params, "--lora", adapters[0])
+	}
+
+	if opts.NumThread > 0 {
+		params = append(params, "--threads", fmt.Sprintf("%d", opts.NumThread))
+	}
+
+	if !opts.F16KV {
+		params = append(params, "--memory-f32")
+	}
+	if opts.UseMLock {
+		params = append(params, "--mlock")
+	}
+	if !opts.UseMMap {
+		params = append(params, "--no-mmap")
+	}
+	if opts.UseNUMA {
+		params = append(params, "--numa")
+	}
+
+	// start the llama.cpp server with a retry in case the port is already in use
+	for try := 0; try < 3; try++ {
+		port := rand.Intn(65535-49152) + 49152 // get a random port in the ephemeral range
+		cmd := exec.Command(
+			runner,
+			append(params, "--port", strconv.Itoa(port))...,
+		)
+		var stderr bytes.Buffer
+		cmd.Stderr = &stderr
+
+		err := cmd.Start()
+		if err != nil {
+			return nil, fmt.Errorf("error starting the external llama.cpp server: %w", err)
+		}
+
+		// relay the logs from the external process
+		go func() {
+			err := cmd.Wait()
+			if err != nil {
+				if err.Error() == "signal: killed" {
+					// this is expected when the server is closed due to loading a new model
+					return
+				}
+				// TODO: what is the specific error when the GPU is not supported?
+				log.Print(stderr.String())
+			}
+		}()
+
+		timeout := time.After(3 * time.Second)
+		ticker := time.NewTicker(100 * time.Millisecond)
+		defer ticker.Stop()
+	next:
+		for {
+			select {
+			case <-timeout:
+				break next
+			case <-ticker.C:
+				resp, err := http.Head(fmt.Sprintf("http://127.0.0.1:%d", port))
+				if err != nil {
+					continue
+				}
+				defer resp.Body.Close()
+				if resp.StatusCode == http.StatusOK {
```

Comment:
It's better to use a looser check, in case the server decides to return some other success code

```suggestion
				if resp.StatusCode < http.StatusBadRequest {
```